File: python/knight-old/number.py
from __future__ import annotations
from knight import Value, Stream, Literal
from typing import Union

class Number(Value):
	@classmethod
	def parse(cls, stream):
		if match := stream.matches(r'\d+'):
			return cls(int(match))

	def __init__(self, data):
		self.data = data

	def __repr__(self):
		return f'Number({self})'

	def run(self):
		return self

	def __str__(self):
		return str(self.run().data)

	def __int__(self):
		return int(self.run().data)

	def __bool__(self):
		return bool(self.run().data)

	def __add__(self, rhs):
		return Number(int(self) + int(rhs))

	def __sub__(self, rhs):
		return Number(int(self) - int(rhs))

	def __mul__(self, rhs):
		return Number(int(self) * int(rhs))

	def __div__(self, rhs):
		return Number(int(self) / int(rhs))

	def __mod__(self, rhs):
		return Number(int(self) % int(rhs))

	def __pow__(self, rhs):
		return Number(int(self) ** int(rhs))

	def __lt__(self, rhs):
		return int(self) < (int(rhs))

	def __eq__(self, rhs):
		return type(self) == type(rhs) and self.data == rhs.data

class Number(Value):
	@classmethod
	def parse(cls, stream):
		if match := stream.matches(r'\d+'):
			return Number(int(match))

File: C#/Number.cs
using System;
using Knight.Ops;
namespace Knight
{
	public class Number : Literal<long>, IAdd, ISub, IMul, IDiv, IMod, IPow, IComparable<IValue>
	{
		public Number(long data) : base(data) {}

		public static IValue Parse(ref string stream) {
			if (!char.IsDigit(stream[0]))
				return null;

			long value = 0;
	
			do {
				value = value * 10 + (stream[0] - '0');

				stream = stream.Substring(1);
			} while(stream != "" && Char.IsDigit(stream[0])) ;

			return new Number(value);
		}

		public override void Dump() => Console.Write($"Number({this})");

		public override bool ToBoolean() => _data != 0;
		public override long ToNumber() => _data;

		public int CompareTo(IValue other) => _data.CompareTo(other.ToNumber());		

		public IValue Add(IValue rhs) => new Number(_data + rhs.ToNumber());
		public IValue Sub(IValue rhs) => new Number(_data - rhs.ToNumber());
		public IValue Mul(IValue rhs) => new Number(_data * rhs.ToNumber());
		public IValue Div(IValue rhs) {
			var rlong = rhs.ToNumber();
			
			if (rlong == 0) {
				throw new RuntimeException("Cannot divide by zero!");
			}

			return new Number(_data / rlong);
		}
		
		public IValue Mod(IValue rhs) {
			var rlong = rhs.ToNumber();
			
			if (rlong == 0) {
				throw new RuntimeException("Cannot modulo by zero!");
			}

			return new Number(_data % rlong);
		}
		
		public IValue Pow(IValue rhs) => new Number((long) Math.Pow(_data, rhs.ToNumber()));
	}
}